WeatherMan Task Submission.

Zainabq-02 opened the pull request.

Zainabq-02 committed:
commit

Zainabq-02 committed:
Update README.md

hammadcui20 commented on .idea/inspectionProfiles/profiles_settings.xml:
The pull request should only consist of code related files only. In our case .py files
The files such as .xml, .iml and pycache are added in gitignore and then commited.

Action Item:
Add the extra directory/files in gitignore and push it to the branch, after than delete the extra files from project directory and push the commit again you file count which is 187 will reduce accordingly.

Zainabq-02 replied:
> The pull request should only consist of code related files only. In our case .py files The files such as .xml, .iml and pycache are added in gitignore and then commited.
> 
> Action Item: Add the extra directory/files in gitignore and push it to the branch, after than delete the extra files from project directory and push the commit again you file count which is 187 will reduce accordingly.

DONE WITH THIS

hammadcui20 commented on data/weather_files/Murree_weather_2004_Aug.txt:
Delete dataset files from the branch and add it in the gitignore as well

hammadcui20 commented on computation.py:
```diff
@@ -0,0 +1,64 @@
+# computation.py
```
Don't need this comment rn

hammadcui20 commented on utils.py:
Remove any empty file

hammadcui20 commented on data_reader.py:
```diff
@@ -0,0 +1,52 @@
+# data_reader.py
```
Same here

hammadcui20 commented on computation.py:
```diff
@@ -0,0 +1,64 @@
+    if not valid_max or not valid_min or not valid_hum:
+        return None
+
+    hottest = max(valid_max, key=lambda r: r.max_temp)
+    coldest = min(valid_min, key=lambda r: r.min_temp)
+    most_humid = max(valid_hum, key=lambda r: r.humidity)
```
`lambda` Why is this used?

Zainabq-02 replied:
> `lambda` Why is this used?

Here lambda is used to tell python what value to use for comparison when finding the maximum or minimum. we can use functions, like define a separate function for each comparison

hammadcui20 commented on data_reader.py:
```diff
@@ -0,0 +1,52 @@
+                                    humidity=humidity
+                                )
+                            )
+                    except Exception:
+                        continue  # skip malformed rows
```
Try avoiding exception handling for now

hammadcui20 commented on models.py:
```diff
@@ -0,0 +1,11 @@
+from dataclasses import dataclass
+from datetime import date
+
+@dataclass
+class WeatherReading:
+    station: str
+    reading_date: date
+    max_temp: int
+    min_temp: int
+    mean_temp: int
+    humidity: int
```
Whats the purpose of using dataclass here? 
Why can't we use simple python class?

Zainabq-02 replied:
> Whats the purpose of using dataclass here? Why can't we use simple python class?

dataclass is used to make a simple class that just stores data.
It automatically creates the __init__, __repr__, and __eq__ methods, so you don’t have to write them yourself. like we can use a normal class, but dataclass makes the code shorter, cleaner, and easier to read.

hammadcui20 commented on parser.py:
```diff
@@ -0,0 +1,105 @@
+                    try:
+                        reading_date = datetime.datetime.strptime(date_str, fmt).date()
+                        break
+                    except ValueError:
```
Avoid exception hanlding here everywhere, try to handle it with if-else cases if possible

hammadcui20 commented on parser.py:
```diff
@@ -0,0 +1,105 @@
+                        mean_temp=mean_temp,
+                        humidity=mean_humidity
+
+                        # ... then append ...
```
Do we need this comment here too?

hammadcui20 reviewed (changes requested):
Changes Needed

Zainabq-02 committed:
Updated .gitignore and removed extra files from git

Zainabq-02 committed:
Removed extra comment, updated daily chart logic

Zainabq-02 committed:
Removed top comment, removed try-except, cleaned data parsing

Zainabq-02 committed:
Removed top comment, removed try-except, cleaned data parsing

Zainabq-02 committed:
Simplified parser, removed comments, avoided try-except

Zainabq-02 commented:
Hi Hammad,

All the requested changes have been made:

Removed extra comments from computation.py and data_reader.py
Deleted empty utils.py
Simplified exception handling and parser logic
Added .xml, .iml, pycache, and dataset files to .gitignore
Only .py files remain in the PR

The PR now reflects these updates. Please review when convenient.

Thank you,
Zainab

On Sun, 31 Aug 2025 at 18:45, Hammad Hussain ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Changes Needed
> ------------------------------
>
> On .idea/inspectionProfiles/profiles_settings.xml
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312465241>
> :
>
> The pull request should only consist of code related files only. In our
> case .py files
> The files such as .xml, .iml and pycache are added in gitignore and then
> commited.
>
> Action Item:
> Add the extra directory/files in gitignore and push it to the branch,
> after than delete the extra files from project directory and push the
> commit again you file count which is 187 will reduce accordingly.
> ------------------------------
>
> On data/weather_files/Murree_weather_2004_Aug.txt
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312465522>
> :
>
> Delete dataset files from the branch and add it in the gitignore as well
> ------------------------------
>
> In computation.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312465725>
> :
>
> > @@ -0,0 +1,64 @@
> +# computation.py
>
> Don't need this comment rn
> ------------------------------
>
> On utils.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312466151>
> :
>
> Remove any empty file
> ------------------------------
>
> In data_reader.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312466362>
> :
>
> > @@ -0,0 +1,52 @@
> +# data_reader.py
>
> Same here
> ------------------------------
>
> In computation.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312466833>
> :
>
> > +    hottest = max(valid_max, key=lambda r: r.max_temp)
> +    coldest = min(valid_min, key=lambda r: r.min_temp)
> +    most_humid = max(valid_hum, key=lambda r: r.humidity)
>
> lambda Why is this used?
> ------------------------------
>
> In data_reader.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312467111>
> :
>
> > +                    except Exception:
> +                        continue  # skip malformed rows
>
> Try avoiding exception handling for now
> ------------------------------
>
> In models.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312467248>
> :
>
> > ***@***.***
> +class WeatherReading:
> +    station: str
> +    reading_date: date
> +    max_temp: int
> +    min_temp: int
> +    mean_temp: int
> +    humidity: int
>
> Whats the purpose of using dataclass here?
> Why can't we use simple python class?
> ------------------------------
>
> In parser.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312467582>
> :
>
> > +
> +            parts = [p.strip() for p in line.split(",")]
> +
> +            # date must exist
> +            if not parts or not parts[0]:
> +                continue
> +
> +            try:
> +                # Flexible date parsing
> +                date_str = parts[0]
> +                reading_date = None
> +                for fmt in ("%Y-%m-%d", "%Y-%m-%d", "%Y-%m-%d"):  # you can add more formats if needed
> +                    try:
> +                        reading_date = datetime.datetime.strptime(date_str, fmt).date()
> +                        break
> +                    except ValueError:
>
> Avoid exception hanlding here everywhere, try to handle it with if-else
> cases if possible
> ------------------------------
>
> In parser.py
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#discussion_r2312467684>
> :
>
> > +
> +                # humidity: choose the column you want to store in WeatherReading.humidity
> +                # currently we store the "Max Humidity" (index 7) because sample outputs used it.
> +                max_humidity = _parse_int(parts[7]) if len(parts) > 7 else None
> +                mean_humidity = _parse_int(parts[8]) if len(parts) > 8 else None
> +
> +                readings.append(
> +                    WeatherReading(
> +                        station=station,
> +                        reading_date=reading_date,
> +                        max_temp=max_temp,
> +                        min_temp=min_temp,
> +                        mean_temp=mean_temp,
> +                        humidity=mean_humidity
> +
> +                        # ... then append ...
>
> Do we need this comment here too?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Zainabq-02/Weatherman-Project/pull/1#pullrequestreview-3171628933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BTWNXMFWTXMUM3W52OSR6G33QL37PAVCNFSM6AAAAACFIDLLSWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCNZRGYZDQOJTGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hammadcui20 commented on computation.py:
```diff
@@ -0,0 +1,62 @@
+    coldest = min(valid_min, key=lambda r: r.min_temp)
+    most_humid = max(valid_hum, key=lambda r: r.humidity)
+
+    return hottest, coldest, most_humid
+
+
+def get_monthly_averages(readings, year, month):
+    monthly = [r for r in readings if r.reading_date.year == year and r.reading_date.month == month]
+
+    max_temps = [r.max_temp for r in monthly if r.max_temp is not None]
+    min_temps = [r.min_temp for r in monthly if r.min_temp is not None]
+    humidities = [r.humidity for r in monthly if r.humidity is not None]
+
+    if not max_temps or not min_temps or not humidities:
+        return None
+
+    avg_high = sum(max_temps) / len(max_temps)
+    avg_low = sum(min_temps) / len(min_temps)
+    avg_humidity = sum(humidities) / len(humidities)
+
+    return avg_high, avg_low, avg_humidity
+
+
+def get_daily_chart(readings, year, month):
```
A good rule of thumb is to leave after function
In case of classes, we usually leave two lines after class in accordance to PEP 8 conventions

hammadcui20 commented on computation.py:
```diff
@@ -0,0 +1,62 @@
+            chart.append((r.reading_date.day, bar_high, bar_low))
+
+    return chart
+
+
+
+#  BONUS FEATURE: Combined daily bar chart
+def get_combined_daily_chart(readings, year, month):
```
Same goes for here too, and in all other files

hammadcui20 commented on data_reader.py:
```diff
@@ -0,0 +1,50 @@
+import os
+from datetime import datetime
+from models import WeatherReading
```
We usually leave one line space if we are importing from custom written modules/files

hammadcui20 commented on weatherman.py:
```diff
@@ -0,0 +1,88 @@
+        if combined_chart:
+            print_combined_daily_chart(combined_chart, year_b, month_b)
+
+    # ---- Print yearly summary at the bottom if user asked monthly but didn't ask yearly ----
```
NIT: No need for this comment

hammadcui20 commented on data_reader.py:
```diff
@@ -0,0 +1,50 @@
+    """
+    readings = []
+
+    for filename in os.listdir(folder_path):
+        if filename.endswith(".txt"):
+            file_path = os.path.join(folder_path, filename)
+
+            with open(file_path, "r") as f:
+                lines = f.readlines()[1:]  # skip header row
+
+                for line in lines:
+                    parts = line.strip().split(",")
+                    if len(parts) < 5:
+                        continue  # skip bad rows
+
+                    station = filename.replace(".txt", "")
+                    reading_date = datetime.strptime(parts[0], "%Y-%m-%d").date()
+
+                    # Only convert if value exists
+                    max_temp = int(parts[1]) if parts[1] else None
```
Try to have as less indentation as we can max limit is of 4, This allows easy maintenance of code as we add new functionalities

hammadcui20 commented on data_reader.py:
Isn't parser and data_reader have same meanings?
Instead of creating of files we can just create a utils file and import all functions from it too

hammadcui20 reviewed (approved):
Feedback related to general PR/code rules, nothing major
